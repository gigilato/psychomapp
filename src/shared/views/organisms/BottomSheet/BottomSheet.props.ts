import { BottomSheetModalProps } from '@gorhom/bottom-sheet'
import { ReactNode } from 'react'

export interface IBottomSheetProps
  extends Omit<BottomSheetModalProps, 'snapPoints' | 'enableDynamicSizing' | 'ref'> {
  properties?: Record<string, string | undefined>
  enableDynamicSnapPoints?: boolean
  snapPoints?: (string | number)[]
  title?: string
  closeOnBackdropPress?: boolean
  children: ReactNode
  modal?: boolean
  containerClassName?: string
  backdropClassName?: string
  backgroundClassName?: string
  handleClassName?: string
}
