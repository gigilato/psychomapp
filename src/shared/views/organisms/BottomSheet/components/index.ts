export * from './BottomSheetBackdrop'
export * from './BottomSheetBackground'
export * from './BottomSheetHandle'
