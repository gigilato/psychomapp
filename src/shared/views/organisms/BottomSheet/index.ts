export * from './BottomSheet'
export * from './BottomSheet.props'
