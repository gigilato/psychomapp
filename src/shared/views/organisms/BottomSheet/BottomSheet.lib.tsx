import BottomSheetDefault, { BottomSheetModal } from '@gorhom/bottom-sheet'
import { ForwardedRef, useCallback, useImperativeHandle, useMemo, useRef, useState } from 'react'
import { LayoutChangeEvent } from 'react-native'

import { spacing } from '$theme'
import { useSafeAreaInsets } from '$views/libs/safeAreaInsets'

import { IBottomSheetProps } from './BottomSheet.props'
import { BottomSheetBackdrop, BottomSheetBackground, BottomSheetHandle } from './components'

export const useBottomSheet = (
  {
    name,
    properties,
    snapPoints = ['40%'],
    index = snapPoints.length - 1,
    closeOnBackdropPress = true,
    enableDynamicSnapPoints = true,
    modal = true,
    ...otherProps
  }: Omit<IBottomSheetProps, 'children'>,
  ref: ForwardedRef<BottomSheetModal>
) => {
  const innerRef = useRef<BottomSheetModal>(null)
  useImperativeHandle(ref, () => ({ ...(innerRef as any).current }))

  const renderBackdrop = useCallback(
    () => (
      <BottomSheetBackdrop
        ID="BottomSheet/backdrop"
        pressBehavior={closeOnBackdropPress ? 'close' : 'none'}
        hitSlop={200}
      />
    ),
    [closeOnBackdropPress]
  )
  const renderBackground = useCallback(() => <BottomSheetBackground />, [])
  const renderHandle = useCallback(() => <BottomSheetHandle />, [])

  const { bottom } = useSafeAreaInsets()

  const [dynamicSnapPoints, setSnapPoints] = useState<(number | string)[]>(snapPoints)
  const onLayout = useCallback(
    (event: LayoutChangeEvent) => {
      if (!enableDynamicSnapPoints) return
      const height = event.nativeEvent.layout.height + bottom + spacing.xl
      setSnapPoints([height, ...snapPoints.slice(1)])
    },
    [snapPoints, bottom, enableDynamicSnapPoints]
  )

  const BottomSheetContainer = useMemo(
    () => (modal ? BottomSheetModal : BottomSheetDefault),
    [modal]
  )

  return {
    ref: innerRef,
    index,
    snapPoints: enableDynamicSnapPoints ? dynamicSnapPoints : snapPoints,
    backdropComponent: renderBackdrop,
    handleComponent: renderHandle,
    backgroundComponent: renderBackground,
    onLayout,
    BottomSheetContainer,
    ...otherProps,
  }
}
