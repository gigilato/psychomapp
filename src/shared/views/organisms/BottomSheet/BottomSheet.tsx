import { BottomSheetModal } from '@gorhom/bottom-sheet'
import React, { forwardRef, memo } from 'react'

import { Box } from '$atoms'

import { useBottomSheet } from './BottomSheet.lib'
import { IBottomSheetProps } from './BottomSheet.props'

export const BottomSheet = memo(
  forwardRef<BottomSheetModal, IBottomSheetProps>(
    ({ children, containerClassName, ...defaultProps }, ref) => {
      const { onLayout, BottomSheetContainer, ...props } = useBottomSheet(defaultProps, ref)
      return (
        <BottomSheetContainer {...props}>
          <Box onLayout={onLayout} className={containerClassName}>
            {children}
          </Box>
        </BottomSheetContainer>
      )
    }
  )
)
