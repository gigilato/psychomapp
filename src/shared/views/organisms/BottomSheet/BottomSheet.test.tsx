import { render } from '@testing-library/react-native'

import { Center, Text } from '$atoms'

import { BottomSheet } from './BottomSheet'

const Content = () => (
  <Center className="h-[40]">
    <Text>Content</Text>
  </Center>
)

describe('BottomSheet', () => {
  it('renders correctly', async () => {
    const component = render(
      <BottomSheet name="DeleteTransfer" modal={false}>
        <Content />
      </BottomSheet>
    )
    expect(component).toMatchSnapshot()
  })

  it('renders without dynamic snap', async () => {
    const component = render(
      <BottomSheet name="DeleteTransfer" modal={false} enableDynamicSnapPoints={false}>
        <Content />
      </BottomSheet>
    )
    expect(component).toMatchSnapshot()
  })

  it('renders with multiple snap points', async () => {
    const component = render(
      <BottomSheet
        name="DeleteTransfer"
        modal={false}
        enableDynamicSnapPoints={false}
        snapPoints={['30%', '70%']}
      >
        <Content />
      </BottomSheet>
    )
    expect(component).toMatchSnapshot()
  })
})
