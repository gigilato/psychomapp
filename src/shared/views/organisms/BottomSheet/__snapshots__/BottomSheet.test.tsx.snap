// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BottomSheet renders correctly 1`] = `
<View
  onLayout={[Function]}
  style={
    [
      [
        undefined,
        undefined,
      ],
    ]
  }
>
  <View
    style={
      [
        {
          "justifyContent": "center",
        },
        {
          "alignItems": "center",
        },
        {
          "height": 40,
        },
        [
          undefined,
          undefined,
        ],
      ]
    }
  >
    <Text
      style={
        [
          {
            "fontFamily": "SpaceGrotesk-Regular",
          },
          {
            "fontSize": 16,
          },
          {
            "lineHeight": 18,
          },
        ]
      }
    >
      Content
    </Text>
  </View>
</View>
`;

exports[`BottomSheet renders with multiple snap points 1`] = `
<View
  onLayout={[Function]}
  style={
    [
      [
        undefined,
        undefined,
      ],
    ]
  }
>
  <View
    style={
      [
        {
          "justifyContent": "center",
        },
        {
          "alignItems": "center",
        },
        {
          "height": 40,
        },
        [
          undefined,
          undefined,
        ],
      ]
    }
  >
    <Text
      style={
        [
          {
            "fontFamily": "SpaceGrotesk-Regular",
          },
          {
            "fontSize": 16,
          },
          {
            "lineHeight": 18,
          },
        ]
      }
    >
      Content
    </Text>
  </View>
</View>
`;

exports[`BottomSheet renders without dynamic snap 1`] = `
<View
  onLayout={[Function]}
  style={
    [
      [
        undefined,
        undefined,
      ],
    ]
  }
>
  <View
    style={
      [
        {
          "justifyContent": "center",
        },
        {
          "alignItems": "center",
        },
        {
          "height": 40,
        },
        [
          undefined,
          undefined,
        ],
      ]
    }
  >
    <Text
      style={
        [
          {
            "fontFamily": "SpaceGrotesk-Regular",
          },
          {
            "fontSize": 16,
          },
          {
            "lineHeight": 18,
          },
        ]
      }
    >
      Content
    </Text>
  </View>
</View>
`;
